File: U1/EXAMEN_SOAP_DOTNET_GR01/01 SERVIDOR/Comercializadora_Soap_DotNet_GR01/Comercializadora_Soap_DotNet_GR01/Services/ProductoService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Comercializadora_Soap_DotNet_GR01.DataAcces.Repositories;
using Comercializadora_Soap_DotNet_GR01.DTOs;
using Comercializadora_Soap_DotNet_GR01.Models;

namespace Comercializadora_Soap_DotNet_GR01.Services
{
    public class ProductoService
    {
        private readonly ProductoRepository _productoRepository;

        public ProductoService()
        {
            _productoRepository = new ProductoRepository();
        }

        public ProductoService(ProductoRepository productoRepository)
        {
            _productoRepository = productoRepository;
        }

        // Crear producto
        public RespuestaDTO CrearProducto(CrearProductoDTO productoDto)
        {
            try
            {
                // Validaciones
                if (string.IsNullOrWhiteSpace(productoDto.Codigo))
                    return new RespuestaDTO { Exito = false, Mensaje = "El código del producto es requerido" };

                if (string.IsNullOrWhiteSpace(productoDto.Nombre))
                    return new RespuestaDTO { Exito = false, Mensaje = "El nombre del producto es requerido" };

                if (productoDto.Precio <= 0)
                    return new RespuestaDTO { Exito = false, Mensaje = "El precio debe ser mayor a cero" };

                if (_productoRepository.ExisteCodigo(productoDto.Codigo))
                    return new RespuestaDTO { Exito = false, Mensaje = $"Ya existe un producto con el código {productoDto.Codigo}" };

                // Mapear DTO a Model (ProductoId y FechaRegistro se autogeneran)
                var producto = new Producto
                {
                    Codigo = productoDto.Codigo,
                    Nombre = productoDto.Nombre,
                    Descripcion = productoDto.Descripcion,
                    Precio = productoDto.Precio,
                    Stock = productoDto.Stock,
                    Categoria = productoDto.Categoria,
                    ImagenUrl = productoDto.ImagenUrl,
                    Estado = "ACTIVO",
                    FechaRegistro = DateTime.Now
                };

                var productoCreado = _productoRepository.Create(producto);

                if (productoCreado == null || productoCreado.ProductoId == 0)
                {
                    return new RespuestaDTO { Exito = false, Mensaje = "Error: El producto no se guardó correctamente en la base de datos" };
                }

                var productoDTO = MapearADTO(productoCreado);

                return new RespuestaDTO
                {
                    Exito = true,
                    Mensaje = $"Producto creado exitosamente con ID: {productoCreado.ProductoId}",
                    Datos = productoDTO
                };
            }
            catch (Exception ex)
            {
                return new RespuestaDTO { Exito = false, Mensaje = $"Error al crear producto: {ex.Message}" };
            }
        }

        // Listar todos los productos
        public List<ProductoDTO> ObtenerProductos()
        {
            var productos = _productoRepository.GetAll();
            return productos.Select(p => MapearADTO(p)).ToList();
        }

        // Obtener producto por ID
        public ProductoDTO ObtenerProductoPorId(int id)
        {
            var producto = _productoRepository.GetById(id);
            return producto != null ? MapearADTO(producto) : null;
        }

        // Obtener producto por código
        public ProductoDTO ObtenerProductoPorCodigo(string codigo)
        {
            var producto = _productoRepository.GetByCodigo(codigo);
            return producto != null ? MapearADTO(producto) : null;
        }

        // Obtener productos por categoría
        public List<ProductoDTO> ObtenerProductosPorCategoria(string categoria)
        {
            var productos = _productoRepository.GetByCategoria(categoria);
            return productos.Select(p => MapearADTO(p)).ToList();
        }

        // Obtener productos por rango de precio
        public List<ProductoDTO> ObtenerProductosPorPrecio(decimal precioMin, decimal precioMax)
        {
            var productos = _productoRepository.GetByPrecioRange(precioMin, precioMax);
            return productos.Select(p => MapearADTO(p)).ToList();
        }

        // Actualizar producto
        public RespuestaDTO ActualizarProducto(ActualizarProductoDTO productoDto)
        {
            try
            {
                // Validaciones
                if (productoDto.ProductoId <= 0)
                    return new RespuestaDTO { Exito = false, Mensaje = "ID de producto inválido" };

                var productoExistente = _productoRepository.GetById(productoDto.ProductoId);
                if (productoExistente == null)
                    return new RespuestaDTO { Exito = false, Mensaje = "Producto no encontrado" };

                if (string.IsNullOrWhiteSpace(productoDto.Nombre))
                    return new RespuestaDTO { Exito = false, Mensaje = "El nombre del producto es requerido" };

                if (productoDto.Precio <= 0)
                    return new RespuestaDTO { Exito = false, Mensaje = "El precio debe ser mayor a cero" };

                if (_productoRepository.ExisteCodigo(productoDto.Codigo, productoDto.ProductoId))
                    return new RespuestaDTO { Exito = false, Mensaje = $"Ya existe otro producto con el código {productoDto.Codigo}" };

                // Actualizar campos (FechaRegistro NO se modifica)
                productoExistente.Codigo = productoDto.Codigo;
                productoExistente.Nombre = productoDto.Nombre;
                productoExistente.Descripcion = productoDto.Descripcion;
                productoExistente.Precio = productoDto.Precio;
                productoExistente.Stock = productoDto.Stock;
                productoExistente.Categoria = productoDto.Categoria;
                productoExistente.ImagenUrl = productoDto.ImagenUrl;
                productoExistente.Estado = productoDto.Estado;

                var productoActualizado = _productoRepository.Update(productoExistente);

                return new RespuestaDTO
                {
                    Exito = true,
                    Mensaje = "Producto actualizado exitosamente",
                    Datos = MapearADTO(productoActualizado)
                };
            }
            catch (Exception ex)
            {
                return new RespuestaDTO { Exito = false, Mensaje = $"Error al actualizar producto: {ex.Message}" };
            }
        }

        // Eliminar producto
        public RespuestaDTO EliminarProducto(int id)
        {
            try
            {
                var producto = _productoRepository.GetById(id);
                if (producto == null)
                    return new RespuestaDTO { Exito = false, Mensaje = "Producto no encontrado" };

                _productoRepository.Delete(id);

                return new RespuestaDTO
                {
                    Exito = true,
                    Mensaje = "Producto eliminado exitosamente"
                };
            }
            catch (Exception ex)
            {
                return new RespuestaDTO { Exito = false, Mensaje = $"Error al eliminar producto: {ex.Message}" };
            }
        }

        // Mapear Model a DTO
        private ProductoDTO MapearADTO(Producto producto)
        {
            return new ProductoDTO
            {
                ProductoId = producto.ProductoId,
                Codigo = producto.Codigo,
                Nombre = producto.Nombre,
                Descripcion = producto.Descripcion,
                Precio = producto.Precio,
                Stock = producto.Stock,
                Categoria = producto.Categoria,
                ImagenUrl = producto.ImagenUrl,
                FechaRegistro = producto.FechaRegistro,
                Estado = producto.Estado
            };
        }
    }
}

File: U2/PROYECTO/04. SERVIDOR/Comercializadora_RestFul_Java_GR01/src/main/java/ec/edu/monster/Services/ProductoService.java
package ec.edu.monster.Services;

import ec.edu.monster.Constants.AppConfig;
import ec.edu.monster.Constants.ErrorMessages;
import ec.edu.monster.DTOs.*;
import ec.edu.monster.Models.Producto;
import ec.edu.monster.Repositories.ProductoRepository;
import ec.edu.monster.Validators.ProductoValidator;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * Servicio para gestión de productos
 */
@Service
@Transactional
public class ProductoService {

    private final ProductoRepository productoRepository;

    public ProductoService(ProductoRepository productoRepository) {
        this.productoRepository = productoRepository;
    }

    /**
     * Crear un nuevo producto
     */
    public RespuestaDTO crearProducto(CrearProductoDTO dto) {
        // Validar usando ProductoValidator
        Optional<String> error = ProductoValidator.validarCrear(dto);
        if (error.isPresent()) {
            return RespuestaDTO.error(error.get());
        }

        // Validar código único
        if (productoRepository.existsByCodigo(dto.getCodigo())) {
            return RespuestaDTO.error(String.format(ErrorMessages.CODIGO_DUPLICADO, dto.getCodigo()));
        }

        // Crear producto
        Producto producto = new Producto();
        producto.setCodigo(dto.getCodigo());
        producto.setNombre(dto.getNombre());
        producto.setDescripcion(dto.getDescripcion());
        producto.setPrecio(dto.getPrecio());
        producto.setStock(dto.getStock() != null ? dto.getStock() : 0);
        producto.setCategoria(dto.getCategoria());
        producto.setImagenUrl(dto.getImagenUrl());
        producto.setEstado(AppConfig.ESTADO_ACTIVO);

        Producto productoCreado = productoRepository.save(producto);

        return RespuestaDTO.exitoso(
                String.format(ErrorMessages.PRODUCTO_CREADO, productoCreado.getProductoId()),
                mapearADTO(productoCreado));
    }

    /**
     * Listar todos los productos activos
     */
    @Transactional(readOnly = true)
    public List<ProductoDTO> obtenerProductos() {
        return productoRepository.findByEstadoOrderByNombreAsc(AppConfig.ESTADO_ACTIVO)
                .stream()
                .map(this::mapearADTO)
                .collect(Collectors.toList());
    }

    /**
     * Obtener producto por ID
     */
    @Transactional(readOnly = true)
    public ProductoDTO obtenerProductoPorId(Integer id) {
        return productoRepository.findById(id)
                .map(this::mapearADTO)
                .orElse(null);
    }

    /**
     * Obtener producto por código
     */
    @Transactional(readOnly = true)
    public ProductoDTO obtenerProductoPorCodigo(String codigo) {
        return productoRepository.findByCodigo(codigo)
                .map(this::mapearADTO)
                .orElse(null);
    }

    /**
     * Obtener productos por categoría
     */
    @Transactional(readOnly = true)
    public List<ProductoDTO> obtenerProductosPorCategoria(String categoria) {
        return productoRepository.findByCategoriaAndEstado(categoria, AppConfig.ESTADO_ACTIVO)
                .stream()
                .map(this::mapearADTO)
                .collect(Collectors.toList());
    }

    /**
     * Obtener productos por rango de precio
     */
    @Transactional(readOnly = true)
    public List<ProductoDTO> obtenerProductosPorPrecio(BigDecimal precioMin, BigDecimal precioMax) {
        return productoRepository.findByPrecioBetween(precioMin, precioMax)
                .stream()
                .map(this::mapearADTO)
                .collect(Collectors.toList());
    }

    /**
     * Actualizar producto (PATCH - solo campos enviados)
     */
    public RespuestaDTO actualizarProducto(Integer id, ActualizarProductoDTO dto) {
        // Validar usando ProductoValidator
        Optional<String> error = ProductoValidator.validarActualizar(dto);
        if (error.isPresent()) {
            return RespuestaDTO.error(error.get());
        }

        // Buscar producto existente
        var productoOpt = productoRepository.findById(id);
        if (productoOpt.isEmpty()) {
            return RespuestaDTO.error(ErrorMessages.PRODUCTO_NO_ENCONTRADO);
        }

        Producto producto = productoOpt.get();

        // Validar código único si se está cambiando
        if (dto.getCodigo() != null && !dto.getCodigo().isBlank()) {
            if (productoRepository.existsByCodigoAndIdNot(dto.getCodigo(), id)) {
                return RespuestaDTO.error(String.format(ErrorMessages.CODIGO_DUPLICADO, dto.getCodigo()));
            }
            producto.setCodigo(dto.getCodigo());
        }

        // Actualizar solo campos enviados (PATCH)
        if (dto.getNombre() != null && !dto.getNombre().isBlank()) {
            producto.setNombre(dto.getNombre());
        }
        if (dto.getDescripcion() != null) {
            producto.setDescripcion(dto.getDescripcion());
        }
        if (dto.getPrecio() != null) {
            producto.setPrecio(dto.getPrecio());
        }
        if (dto.getStock() != null) {
            producto.setStock(dto.getStock());
        }
        if (dto.getCategoria() != null) {
            producto.setCategoria(dto.getCategoria());
        }
        if (dto.getImagenUrl() != null) {
            producto.setImagenUrl(dto.getImagenUrl());
        }
        if (dto.getEstado() != null && !dto.getEstado().isBlank()) {
            producto.setEstado(dto.getEstado());
        }

        Producto productoActualizado = productoRepository.save(producto);

        return RespuestaDTO.exitoso(
                ErrorMessages.PRODUCTO_ACTUALIZADO,
                mapearADTO(productoActualizado));
    }

    /**
     * Eliminar producto
     */
    public RespuestaDTO eliminarProducto(Integer id) {
        var productoOpt = productoRepository.findById(id);
        if (productoOpt.isEmpty()) {
            return RespuestaDTO.error(ErrorMessages.PRODUCTO_NO_ENCONTRADO);
        }

        productoRepository.deleteById(id);
        return RespuestaDTO.exitoso(ErrorMessages.PRODUCTO_ELIMINADO);
    }

    // Mapear entidad a DTO
    private ProductoDTO mapearADTO(Producto producto) {
        return new ProductoDTO(
                producto.getProductoId(),
                producto.getCodigo(),
                producto.getNombre(),
                producto.getDescripcion(),
                producto.getPrecio(),
                producto.getStock(),
                producto.getCategoria(),
                producto.getImagenUrl(),
                producto.getFechaRegistro(),
                producto.getEstado());
    }
}
